src/never.rs

```diff
@@ -0,0 +1,91 @@
+use {Future, Poll};
+
+/// A future which is never resolved.
+///
+/// This future can be created with the `empty` function.
+#[derive(Copy, Clone)]
+pub struct Never {}
+
+/// Creates a future which never resolves, representing a computation that never
+/// finishes.
+///
+/// The returned future will never resolve with a success but is still
+/// susceptible to cancellation. That is, if a callback is scheduled on the
+/// returned future, it is only run once the future is dropped (canceled).
+pub fn never<T, E>() -> impl Future<Item = T, Error = E> + Send + 'static {
```

Comment:
Your confusion is definitely understandable-- lots of things have gotten in the way of this being the simple change I proposed. There's been lots of clarification on what can and can't be done with the `!` type and with default type parameters.

The starting goal was to make `Never`, `Finished`, and `Failed` easier to understand and use. In order to do this, I proposed that the type of the unreachable variants be changed to `!`. This change would make `Never`, `Finished`, and `Failed` easier to understand by clearly expressing (via types) that some of their variants can never exist. The change should also make them easier to use, as listing explicit types should no longer be necessary, and easy conversion between types should be possible.

The extra methods on `!`-based `Future`s should be added because they allow for conversion of an existing `!`, e.g. `Future<T, !>` to `Future<T, E>`.

That said, unless we can get default parameters on functions I think we're stuck. [This](https://is.gd/UJDwOk) is what we want, but functions can't have default type parameters. I've tried a number of different iterations on this, but I always get stuck with this same error. @nikomatsakis, is there any chance you know how we can get a `never` function that defaults to `Never<!, !>`, but can also produce a `Never<T, E>`? I'm mainly just really confused about why default type parameters aren't available for functions.